src/extra/gfx/camera.cpp

```diff
@@ -0,0 +1,187 @@
+          break;
+        case SDL_KeyCode::SDLK_s:
+          _input.keys.backward = event.key.type == SDL_KEYDOWN;
+          break;
+        case SDL_KeyCode::SDLK_d:
+          _input.keys.right = event.key.type == SDL_KEYDOWN;
+          break;
+        }
+        break;
+      case SDL_MOUSEMOTION:
+        _input.look.cursorPosition = float2(event.motion.x, event.motion.y);
+        break;
+      case SDL_MOUSEBUTTONDOWN:
+      case SDL_MOUSEBUTTONUP:
+        _input.look.cursorPosition = float2(event.button.x, event.button.y);
+        switch (event.button.button) {
+        case SDL_BUTTON_LEFT:
+          _input.look.primary = event.button.type == SDL_MOUSEBUTTONDOWN;
+          break;
+        case SDL_BUTTON_RIGHT:
+          _input.look.secondary = event.button.type == SDL_MOUSEBUTTONDOWN;
+          break;
+        case SDL_BUTTON_MIDDLE:
+          _input.look.tertiary = event.button.type == SDL_MOUSEBUTTONDOWN;
+          break;
+        }
+        break;
+      case SDL_MOUSEWHEEL:
+        _input.zoom += float(event.wheel.preciseY);
+        break;
+      }
+    }
+  }
+
+  template <typename T> T getParamVarOrDefault(ParamVar &paramVar, const T &_default) {
+    Var var(paramVar.get());
+    return var.isNone() ? _default : (T)var;
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    updateInput();
+
+    const float mouseBaseFactor = 2000.0f;
+
+    float flySpeed = getParamVarOrDefault(_flySpeed, 1.0f) * 1.0f;
+    float lookSpeed = getParamVarOrDefault(_lookSpeed, 1.0f) / mouseBaseFactor * 0.7f * gfx::pi2;
+    float scrollSpeed = getParamVarOrDefault(_scrollSpeed, 1.0f);
+    float panSpeed = getParamVarOrDefault(_panSpeed, 1.0f) / mouseBaseFactor * 2.0f;
+
+    float4x4 viewMatrix = (Mat4)input;
+    float4x4 invViewMatrix = linalg::inverse(viewMatrix);
+
+    float3 cameraTranslation;
+    float3 cameraScale;
+    float3x3 cameraRotationMatrix;
+    gfx::decomposeTRS(invViewMatrix, cameraTranslation, cameraScale, cameraRotationMatrix);
+    float4 cameraRotation = linalg::rotation_quat(cameraRotationMatrix);
+
+    float3 cameraX = linalg::qxdir(cameraRotation);
+    float3 cameraY = linalg::qydir(cameraRotation);
+    float3 cameraZ = linalg::qzdir(cameraRotation);
+
+    float2 cursorDelta = _input.look.cursorPosition - _input.look.prevCursorPosition;
+
+    if (_input.look.secondary) {
+      // Apply look rotation
+      float4 rotY = linalg::rotation_quat(cameraY, -cursorDelta.x * lookSpeed);
+      float4 rotX = linalg::rotation_quat(cameraX, -cursorDelta.y * lookSpeed);
+      cameraRotation = linalg::qmul(rotX, cameraRotation);
+      cameraRotation = linalg::qmul(rotY, cameraRotation);
+
+      // Fly keys
+      float2 wasdVec{};
+      if (_input.keys.left)
+        wasdVec.x += -1.0f;
+      if (_input.keys.right)
+        wasdVec.x += 1.0f;
+      if (_input.keys.forward)
+        wasdVec.y += 1.0f;
+      if (_input.keys.backward)
+        wasdVec.y += -1.0f;
+
+      // TODO
```

Comment:
Unclear comment - explain what needs to be done.

Reply:
Fixed the issue while moving some things around here